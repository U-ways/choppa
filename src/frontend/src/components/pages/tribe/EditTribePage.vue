<template>
  <FixedWidthWithNavbarTemplate css="pt-3">
    <div class="edit-tribe mb-5">
      <div class="edit-tribe__header text-center">
        Edit Tribe
      </div>
      <div class="edit-tribe__body">
        <div class="row">
          <div class="col-md-3 edit-tribe__subheading pb-2 pb-md-0">
            Tribe
          </div>
          <div class="col-md-9 p-0 px-3 pl-md-0">
            <div class="form-group">
              <label for="tribe-name">Tribe Name</label>
              <input class="form-control" id="tribe-name" v-model="tribe.name">
            </div>
          </div>
        </div>
        <div class="edit-tribe__section">
          <div class="row">
            <div class="col-md-3 edit-tribe__subheading pb-2 pb-md-0">
              Chapters
            </div>
            <div class="col-md-9">
              <div class="mb-4" v-for="chapter in tribe.allDistinctChapters()" :key="chapter.id">
                <div class="row">
                  <div class="col-8 p-0 pl-3 pl-md-0">
                    <label :for="chapterInputId(chapter.id)">Chapter Name</label>
                  </div>
                </div>
                <div class="row">
                  <div class="col-8 p-0 pl-3 pl-md-0">
                    <input class="form-control"
                           :id="chapterInputId(chapter.id)"
                           :value="chapter.name"
                           @change="onChapterNameChanged(chapter, $event)"
                    >
                  </div>
                  <div class="col-2 px-1 px-md-3">
                    <ColourSquareMolecule
                      :startingColour="chapter.colour"
                      @colourChanged="onChapterColourChanged"
                      :returnEvent="{ chapter: chapter }"
                    />
                  </div>
                  <div class="col-2 p-0">
                    <button type="button" class="btn btn-outline-dark" @click="onChapterDeleted(chapter)">
                      <font-awesome-icon icon="trash"/>
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="edit-tribe__section">
          <div class="row">
            <div class="col-12 col-md-3 edit-tribe__subheading pb-2 pb-md-0">
              Squads
            </div>
            <div class="col-12 col-md-9">
              <div class="mb-4" v-for="squad in tribe.squads" :key="squad.id">
                <div class="row">
                  <div class="col-8 p-0 pl-3 pl-md-0">
                    <label :for="nameInputId(squad.id)">Squad Name</label>
                  </div>
                </div>
                <div class="row">
                  <div class="col-8 p-0 pl-3 pl-md-0">
                    <input class="form-control"
                           :id="nameInputId(squad.id)"
                           v-model="tribe.findSquadById(squad.id).name"
                    >
                  </div>
                  <div class="col-2 px-1 px-md-3">
                    <ColourSquareMolecule
                      :startingColour="tribe.findSquadById(squad.id).colour"
                      @colourChanged="onSquadColourChanged"
                      :returnEvent="{ squadId: squad.id }"
                    />
                  </div>
                  <div class="col-2 p-0">
                    <button type="button" class="btn btn-outline-dark" @click="deleteSquad(squad.id)">
                      <font-awesome-icon icon="trash"/>
                    </button>
                  </div>
                </div>
                <div class="row mt-md-3 justify-content-center">
                  <div class="col-12 p-md-0 text-secondary pt-3 d-flex align-items-center">
                    <span>Members</span>
                  </div>
                </div>
                <div class="row mt-2">
                  <div class="col-12 p-md-0 bg-white mb-2" v-for="member in squad.members" :key="member.id">
                    <EditTribeMemberRow :member="member"
                                        :possible-chapters="tribe.allDistinctChapters()"
                                        :possible-squads="tribe.squads"
                                        :current-squad="squad"
                                        @expanded="onMemberRowExpanded"
                                        @collapsed="onMemberRowCollapsed"
                    />
                  </div>
                </div>
                <div class="row mt-2">
                  <div class="col-12 pl-3 pl-md-0">
                    <button type="button" class="btn btn-sm btn-outline-dark" @click="addNewMember(squad)">
                      <font-awesome-icon icon="plus"/>
                      Add New Member
                    </button>
                  </div>
                </div>
              </div>
              <div class="row mt-2">
                <div class="col-12 pl-3 pl-md-0">
                  <button type="button" class="btn btn-outline-dark" @click="addNewSquad">
                    <font-awesome-icon icon="plus"/>
                    Add New Squad
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div class="edit-tribe__section">
          <div class="row">
            <div class="col-12 col-md-3 pb-2 pb-md-0">
              <p class="edit-tribe__subheading">Rotation</p>
              <p class="text-muted">You can manually trigger a rotation here.</p>
            </div>
            <div class="col-12 col-md-9 d-flex align-items-center">
              <button class="btn btn-block btn-secondary">Trigger Rotation</button>
            </div>
          </div>
        </div>
        <div class="edit-tribe__section">
          <div class="row">
            <div class="col-12 col-md-3 pb-2 pb-md-0">
              <p class="edit-tribe__subheading">Save changes</p>
              <p class="text-muted">Save your changes before leaving this page.</p>
            </div>
            <div class="col-12 col-md-9 d-flex align-items-center">
              <button class="btn btn-block btn-primary">Save</button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </FixedWidthWithNavbarTemplate>
</template>

<script>
import FixedWidthWithNavbarTemplate from "@/components/templates/FixedWidthWithNavbarTemplate";
import ColourSquareMolecule from "@/components/molecules/ColourSquareMolecule";
import EditTribeMemberRow from "@/components/molecules/EditTribeMemberRow";
import { v4 as uuidv4 } from "uuid";
import Member from "@/data/types/Member";
import Squad from "@/data/types/Squad";

export default {
  name: "EditTribePage",
  components: {
    EditTribeMemberRow,
    ColourSquareMolecule,
    FixedWidthWithNavbarTemplate,
  },
  data() {
    return {
      tribe: this.$root.$data.testTribeOne,
      currentlyExpandedMemberRow: undefined,
    };
  },
  methods: {
    nameInputId(squadId) {
      return `squad-name-${squadId}`;
    },
    chapterInputId(chapterId) {
      return `chapter-name-${chapterId}`;
    },
    memberDropdownSettingsId(memberId) {
      return `member-dropdown-setting-${memberId}`;
    },
    deleteSquad(squadId) {
      this.tribe.removeSquadById(squadId);
    },
    onSquadColourChanged(eventData) {
      this.tribe.findSquadById(eventData.squadId).colour = eventData.colour.hex;
    },
    onChapterColourChanged(eventData) {
      this.tribe.updateChapter(eventData.chapter.id, eventData.chapter.name, eventData.colour.hex);
    },
    onChapterNameChanged(chapter, eventData) {
      this.tribe.updateChapter(chapter.id, eventData.target.value, chapter.colour);
    },
    onMemberRowExpanded(memberRow) {
      if (this.currentlyExpandedMemberRow) {
        this.currentlyExpandedMemberRow.silentCollapse();
      }

      this.currentlyExpandedMemberRow = memberRow;
    },
    onMemberRowCollapsed(memberRow) {
      if (this.currentlyExpandedMemberRow === memberRow) {
        this.currentlyExpandedMemberRow = null;
      }
    },
    onChapterDeleted(chapter) {
      console.log(chapter);
    },
    addNewMember(squad) {
      squad.addMember(new Member(uuidv4(), "New Member", undefined));
      // squad.addMember(new Member(uuidv4(), "", undefined));
    },
    addNewSquad() {
      this.tribe.addSquad(new Squad(uuidv4(), "New Squad", "#ff00ff", []));
    },
  },
};
</script>

<style scoped lang="scss">
@import "src/assets/scss/colours";
@import "src/assets/scss/typography";

.edit-tribe {
  background: $white;
  border: 1px solid $white-border;
  padding: 1rem;

  &__header {
    font-size: 1.8rem;
    font-family: $font-family-emphasis;
  }

  &__subheading {
    font-family: $font-family-body;
    font-weight: $font-weight-body-bold;
  }

  &__body {
    margin-top: 1rem;
  }

  &__section {
    border-top: 1px solid $white-border;
    margin-top: 1rem;
    padding-top: 0.3rem;
  }
}

</style>
