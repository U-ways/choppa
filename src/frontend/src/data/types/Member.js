export default class Member {
  constructor(id, name, chapter) {
    this._id = id;
    this._name = name;
    this._chapter = chapter;
  }

  get id() {
    return this._id;
  }

  get name() {
    return this._name;
  }

  set name(newName) {
    this._name = newName;
  }

  get chapter() {
    return this._chapter;
  }
}
