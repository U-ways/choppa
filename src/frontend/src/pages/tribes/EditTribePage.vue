<template>
  <StandardPageTemplate>
    <template v-slot:page-header v-if="tribe">
      <div class="flex flex-row place-content-between place-items-center">
        <div class="text-3xl font-normal truncate">
          <span class="hidden sm:inline">Tribe </span>
          <span class="font-bold truncate">{{tribeNameHeader}}</span>
        </div>
        <div class="flex-shrink-0">
          <StyledButton type="link" variant="secondary" css="px-2 sm:pr-5 sm:pl-4"
                        :link="`/${tribe.id}`">
            <div class="inline sm:pr-1">
              <font-awesome-icon icon="eye"/>
            </div>
            <div class="hidden sm:inline">View</div>
          </StyledButton>
        </div>
      </div>
    </template>
    <template v-slot:fixed-width v-if="tribe">
      <div class="px-3 py-5"> <!-- my-5 bg-white rounded-md shadow-md -->
        <section>
          <FormHeader>
            <template v-slot:heading>Tribe Settings</template>
            <template v-slot:subheading>Let's get started by filling in the information about the Tribe.</template>
          </FormHeader>
          <div class="pl-4 flex flex-col gap-2 mt-4">
            <StandardInputWithLabel id="tribe-name"
                                    label-text="Tribe Name"
                                    place-holder="Untitled"
                                    v-bind:value="tribe.name"
                                    v-on:input="tribe.name = $event.target.value"
                                    :validator="$v.tribe.name"
                                    :max-length="100"
            />
            <div class="self-end">
              <StyledButton type="button" variant="primary" css="px-2 pr-5 pl-4" @click="save">
                <font-awesome-icon icon="check"/>
                Save
              </StyledButton>
            </div>
          </div>
        </section>
        <section class="mt-5">
          <FormHeader>
            <template v-slot:heading>Squad Settings</template>
            <template v-slot:subheading>Now lets add some Squads.</template>
          </FormHeader>
          <div class="pl-4 pt-3">
            <div v-if="tribe.squads.length > 0" class="flex flex-col gap-2">
              <div class="grid grid-cols-1 md:grid-cols-2 gap-2">
                <router-link v-for="squad in tribe.squads" v-bind:key="squad.id" :to="`/${squad.id}`"
                             class="hover:ring-2 focus:ring-2 focus:outline-none ring-choppa-two rounded-sm">
                  <SquadCard :squad="squad"
                             :show-header-only="true"
                             :member-show-count="6"
                             :headerLinkLocation="squad.id"
                  />
                </router-link>
              </div>
              <div class="self-end">
                <StyledButton type="button" variant="secondary" css="px-2 pr-5 pl-4">
                  <font-awesome-icon icon="plus"/>
                  New Squad
                </StyledButton>
              </div>
            </div>
            <NoSquadsToShowAlert v-else />
          </div>
        </section>
      </div>
    </template>
  </StandardPageTemplate>
</template>

<script>
import StandardPageTemplate from "@/components/templates/StandardPageTemplate";
import StyledButton from "@/components/atoms/buttons/StyledButton";
import FormHeader from "@/components/forms/FormHeader";
import StandardInputWithLabel from "@/components/forms/StandardInputWithLabel";
import SquadCard from "@/components/squads/SquadCard";
import { getTribe, saveTribe } from "@/config/api/tribe.api";
import NoSquadsToShowAlert from "@/components/tribes/NoSquadsToShowAlert";
import { required, minLength, maxLength } from "vuelidate/lib/validators";
import { mapActions } from "vuex";
import { toastVariants } from "@/enums/toastVariants";
import ToastData from "@/models/toastData";

export default {
  name: "ViewTribePage",
  components: {
    NoSquadsToShowAlert,
    SquadCard,
    StandardInputWithLabel,
    FormHeader,
    StandardPageTemplate,
    StyledButton,
  },
  data() {
    return {
      tribe: undefined,
    };
  },
  validations: {
    tribe: {
      name: {
        required,
        minLength: minLength(3),
        maxLength: maxLength(100),
      },
    },
  },
  async mounted() {
    try {
      this.tribe = await getTribe({ id: this.$route.params.id });
    } catch (error) {
      await this.$router.replace("/not-found");
    }
  },
  computed: {
    tribeNameHeader() {
      return this.tribe.name ? this.tribe.name : "Untitled";
    },
  },
  methods: {
    ...mapActions(["newToast"]),
    save() {
      if (this.$v.$invalid) {
        return;
      }

      try {
        saveTribe({ tribe: this.tribe });
        this.$router.replace(`/${this.tribe.id}`);
        this.newToast(new ToastData({
          variant: toastVariants.SUCCESS,
          message: `Tribe ${this.tribe.name} has been updated`,
        }));
      } catch (error) {
        this.newToast(new ToastData({ variant: toastVariants.ERROR, message: "Save failed, please try again later" }));
        throw error;
      }
    },
  },
};
</script>
