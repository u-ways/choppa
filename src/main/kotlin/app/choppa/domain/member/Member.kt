package app.choppa.domain.member

import app.choppa.domain.base.BaseModel
import app.choppa.domain.chapter.Chapter
import app.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import app.choppa.domain.squad.Squad
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @Column(name = "active", columnDefinition = "BOOLEAN")
    val active: Boolean = true,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=${chapter.name})"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
