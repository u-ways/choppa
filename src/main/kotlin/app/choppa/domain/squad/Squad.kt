package app.choppa.domain.squad

import app.choppa.domain.account.Account
import app.choppa.domain.account.Account.Companion.PLACEHOLDER_ACCOUNT
import app.choppa.domain.base.BaseModel
import app.choppa.domain.member.Member
import app.choppa.domain.member.Member.Companion.NO_MEMBERS
import app.choppa.domain.tribe.Tribe
import app.choppa.domain.tribe.Tribe.Companion.PLACEHOLDER_TRIBE
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*
import javax.persistence.FetchType.EAGER

@Entity
@Table(name = "squad")
@JsonSerialize(using = SquadSerializer::class)
@JsonDeserialize(using = SquadDeserializer::class)
data class Squad(
    @Id
    @Column(name = "squad_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String,

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int,

    // TODO(U-ways) we might be able to lazy fetch this (default is eager)
    @ManyToOne
    @JoinColumn(name = "tribe", referencedColumnName = "tribe_id")
    val tribe: Tribe,

    @ManyToMany(fetch = EAGER)
    @JoinTable(
        name = "squad_current_members",
        joinColumns = [JoinColumn(name = "squad_id", referencedColumnName = "squad_id")],
        inverseJoinColumns = [JoinColumn(name = "member_id", referencedColumnName = "member_id")]
    )
    val members: MutableList<Member> = NO_MEMBERS,

    @ManyToOne
    @JoinColumn(name = "account_id", referencedColumnName = "account_id")
    override val account: Account,
) : BaseModel {
    override fun toString() = "Squad(id=$id, name=$name, tribe=$tribe)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Squad
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_SQUADS: MutableList<Squad>
            get() = mutableListOf()
        val PLACEHOLDER_SQUAD: Squad
            get() = Squad(name = "", color = 0, tribe = PLACEHOLDER_TRIBE, account = PLACEHOLDER_ACCOUNT)
    }
}
