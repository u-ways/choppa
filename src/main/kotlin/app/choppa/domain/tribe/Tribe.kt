package app.choppa.domain.tribe

import app.choppa.domain.base.BaseModel
import app.choppa.domain.squad.Squad
import app.choppa.utils.Color.Companion.GREY
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.hibernate.annotations.BatchSize
import org.hibernate.annotations.Fetch
import org.hibernate.annotations.FetchMode.SELECT
import org.hibernate.annotations.GenericGenerator
import java.util.*
import java.util.UUID.randomUUID
import javax.persistence.*
import javax.persistence.FetchType.EAGER

@Entity
@Table(name = "tribe")
@JsonSerialize(using = TribeSerializer::class)
@JsonDeserialize(using = TribeDeserializer::class)
data class Tribe(
    @Id
    @Column(name = "tribe_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "TR-$id".substring(0, 15),

    @Column(name = "color", columnDefinition = "INTEGER")
    val color: Int = GREY,

    @Fetch(SELECT)
    @BatchSize(size = 10)
    @OneToMany(mappedBy = "tribe", fetch = EAGER)
    val squads: MutableList<Squad> = mutableListOf(),
) : BaseModel {
    override fun toString() = "Tribe(id=$id, name=$name)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Tribe
        if (id != other.id) return false
        return true
    }

    companion object {
        val UNASSIGNED_TRIBE = Tribe(UUID.fromString("00000000-0000-0000-0000-000000000000"), "Unassigned Squad")
    }
}
