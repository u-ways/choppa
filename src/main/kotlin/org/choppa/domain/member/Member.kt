package org.choppa.domain.member

import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import org.choppa.domain.base.BaseModel
import org.choppa.domain.chapter.Chapter
import org.choppa.domain.chapter.Chapter.Companion.UNASSIGNED_ROLE
import org.choppa.domain.history.History
import org.choppa.domain.squad.Squad
import org.hibernate.annotations.GenericGenerator
import java.util.UUID
import java.util.UUID.randomUUID
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.ManyToMany
import javax.persistence.ManyToOne
import javax.persistence.OneToMany
import javax.persistence.Table

@Entity
@Table(name = "member")
@JsonSerialize(using = MemberSerializer::class)
@JsonDeserialize(using = MemberDeserializer::class)
data class Member(
    @Id
    @Column(name = "member_id", columnDefinition = "uuid")
    @GenericGenerator(name = "UUID", strategy = "org.hibernate.id.UUIDGenerator")
    override val id: UUID = randomUUID(),

    @Column(name = "name", columnDefinition = "VARCHAR(100)", nullable = false)
    val name: String = "ME-$id".substring(0, 15),

    @ManyToOne
    @JoinColumn(name = "chapter", referencedColumnName = "chapter_id")
    val chapter: Chapter = UNASSIGNED_ROLE,

    @ManyToMany(mappedBy = "members")
    val squads: MutableList<Squad> = mutableListOf(),

    @OneToMany(mappedBy = "member")
    val history: MutableList<History> = mutableListOf()
) : BaseModel {
    override fun toString() = "Member(id=$id, name=$name, chapter=$chapter)"

    override fun hashCode(): Int = id.hashCode()

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false
        other as Member
        if (id != other.id) return false
        return true
    }

    companion object {
        val NO_MEMBERS get() = mutableListOf<Member>()
    }
}
