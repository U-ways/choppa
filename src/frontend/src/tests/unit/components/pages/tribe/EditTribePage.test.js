import { createLocalVue, describe, expect, it, shallowMount } from "@/tests/unit/test-imports";
import VueRouter from "vue-router";
import EditTribePage from "@/components/pages/tribe/EditTribePage";
import Member from "@/data/types/member";
import Squad from "@/data/types/squad";
import Tribe from "@/data/types/tribe";

const tribe = new Tribe(4096, "Mr Tribe", [
  new Squad(1024, "Mr SquadOne", "#FFFFFF", [
    new Member(1, "Johnny"),
    new Member(2, "Robert"),
    new Member(3, "Peter"),
    new Member(10, "Steve"),
    new Member(11, "Johnny"),
  ]),
  new Squad(2048, "Mr SquadTwo", "#FFFFFF", [
    new Member(1, "Johnny"),
    new Member(2, "Robert"),
    new Member(3, "Peter"),
  ]),
]);

function mountComponent() {
  const localVue = createLocalVue();
  localVue.use(VueRouter);
  const router = new VueRouter();

  return shallowMount(EditTribePage, {
    localVue,
    router,
    data() {
      return {
        tribe,
      };
    },
    stubs: {
      "font-awesome-icon": true,
    },
  });
}

describe("Edit Tribe Page test", () => {
  it("nameInputId", () => {
    const wrapper = mountComponent();
    expect(wrapper.vm.nameInputId(5)).toBe("squad-name-5");
  });

  it("deleteSquad", async () => {
    const mock = jest.spyOn(Tribe.prototype, "removeSquadById");

    const wrapper = mountComponent();
    wrapper.vm.deleteSquad(5);
    expect(mock).toHaveBeenCalledTimes(1);
    expect(mock).toHaveBeenCalledWith(5);

    jest.restoreAllMocks();
  });

  it("onSquadColourChanged", async () => {
    const squad = new Squad("250", "Test Colour Squad", "old colour", []);
    jest.spyOn(Tribe.prototype, "findSquadById").mockReturnValue(squad);

    const wrapper = mountComponent();
    wrapper.vm.onSquadColourChanged({
      squadId: 5,
      colour: {
        hex: "new colour",
      },
    });

    expect(squad.colour).toEqual("new colour");
    jest.restoreAllMocks();
  });

  it("Renders correctly", async () => {
    const wrapper = mountComponent();
    expect(wrapper.html()).toMatchSnapshot();
  });
});
