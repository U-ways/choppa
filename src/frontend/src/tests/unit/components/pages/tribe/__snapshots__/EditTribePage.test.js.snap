// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Edit Tribe Page test Renders correctly 1`] = `
"<fixedwidthwithnavbartemplate-stub css=\\"pt-3\\">
  <div class=\\"edit-tribe mb-5\\">
    <div class=\\"edit-tribe__header text-center\\">
      Edit Tribe
    </div>
    <div class=\\"edit-tribe__body\\">
      <div class=\\"row\\">
        <div class=\\"col-md-3 edit-tribe__subheading pb-2 pb-md-0\\">
          Tribe
        </div>
        <div class=\\"col-md-9 p-0 px-3 pl-md-0\\">
          <div class=\\"form-group\\"><label for=\\"tribe-name\\">Tribe Name</label> <input id=\\"tribe-name\\" class=\\"form-control\\"></div>
        </div>
      </div>
      <div class=\\"edit-tribe__section\\">
        <div class=\\"row\\">
          <div class=\\"col-md-3 edit-tribe__subheading pb-2 pb-md-0\\">
            Chapters
          </div>
          <div class=\\"col-md-9\\">
            <div class=\\"row mt-2\\">
              <div class=\\"col-12 pl-3 pl-md-0\\"><button type=\\"button\\" class=\\"btn btn-outline-dark\\">
                  <font-awesome-icon-stub icon=\\"plus\\"></font-awesome-icon-stub>
                  Add New Chapter
                </button></div>
            </div>
          </div>
        </div>
      </div>
      <div class=\\"edit-tribe__section\\">
        <div class=\\"row\\">
          <div class=\\"col-12 col-md-3 edit-tribe__subheading pb-2 pb-md-0\\">
            Squads
          </div>
          <div class=\\"col-12 col-md-9\\">
            <div class=\\"mb-4\\">
              <div class=\\"row\\">
                <div class=\\"col-8 p-0 pl-3 pl-md-0\\"><label for=\\"squad-name-1024\\">Squad Name</label></div>
              </div>
              <div class=\\"row\\">
                <div class=\\"col-8 p-0 pl-3 pl-md-0\\"><input id=\\"squad-name-1024\\" class=\\"form-control\\"></div>
                <div class=\\"col-2 px-1 px-md-3\\">
                  <coloursquaremolecule-stub startingcolour=\\"#FFFFFF\\" returnevent=\\"[object Object]\\"></coloursquaremolecule-stub>
                </div>
                <div class=\\"col-2 p-0\\"><button type=\\"button\\" class=\\"btn btn-outline-dark\\">
                    <font-awesome-icon-stub icon=\\"trash\\"></font-awesome-icon-stub>
                  </button></div>
              </div>
              <div class=\\"row mt-md-3 justify-content-center\\">
                <div class=\\"col-12 p-md-0 text-secondary pt-3 d-flex align-items-center\\"><span>Members</span></div>
              </div>
              <div class=\\"row mt-2\\">
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
              </div>
              <div class=\\"row mt-2\\">
                <div class=\\"col-12 pl-3 pl-md-0\\"><button type=\\"button\\" class=\\"btn btn-sm btn-outline-dark\\">
                    <font-awesome-icon-stub icon=\\"plus\\"></font-awesome-icon-stub>
                    Add New Member
                  </button></div>
              </div>
            </div>
            <div class=\\"mb-4\\">
              <div class=\\"row\\">
                <div class=\\"col-8 p-0 pl-3 pl-md-0\\"><label for=\\"squad-name-2048\\">Squad Name</label></div>
              </div>
              <div class=\\"row\\">
                <div class=\\"col-8 p-0 pl-3 pl-md-0\\"><input id=\\"squad-name-2048\\" class=\\"form-control\\"></div>
                <div class=\\"col-2 px-1 px-md-3\\">
                  <coloursquaremolecule-stub startingcolour=\\"#FFFFFF\\" returnevent=\\"[object Object]\\"></coloursquaremolecule-stub>
                </div>
                <div class=\\"col-2 p-0\\"><button type=\\"button\\" class=\\"btn btn-outline-dark\\">
                    <font-awesome-icon-stub icon=\\"trash\\"></font-awesome-icon-stub>
                  </button></div>
              </div>
              <div class=\\"row mt-md-3 justify-content-center\\">
                <div class=\\"col-12 p-md-0 text-secondary pt-3 d-flex align-items-center\\"><span>Members</span></div>
              </div>
              <div class=\\"row mt-2\\">
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
                <div class=\\"col-12 p-md-0 bg-white mb-2\\">
                  <edittribememberrow-stub member=\\"[object Object]\\" possiblechapters=\\"\\" possiblesquads=\\"[object Object],[object Object]\\" currentsquad=\\"[object Object]\\"></edittribememberrow-stub>
                </div>
              </div>
              <div class=\\"row mt-2\\">
                <div class=\\"col-12 pl-3 pl-md-0\\"><button type=\\"button\\" class=\\"btn btn-sm btn-outline-dark\\">
                    <font-awesome-icon-stub icon=\\"plus\\"></font-awesome-icon-stub>
                    Add New Member
                  </button></div>
              </div>
            </div>
            <div class=\\"row mt-2\\">
              <div class=\\"col-12 pl-3 pl-md-0\\"><button type=\\"button\\" class=\\"btn btn-outline-dark\\">
                  <font-awesome-icon-stub icon=\\"plus\\"></font-awesome-icon-stub>
                  Add New Squad
                </button></div>
            </div>
          </div>
        </div>
      </div>
      <div class=\\"edit-tribe__section\\">
        <div class=\\"row\\">
          <div class=\\"col-12 col-md-3 pb-2 pb-md-0\\">
            <p class=\\"edit-tribe__subheading\\">Rotation</p>
            <p class=\\"text-muted\\">You can manually trigger a rotation here.</p>
          </div>
          <div class=\\"col-12 col-md-9 d-flex align-items-center\\"><button class=\\"btn btn-block btn-secondary\\">Trigger Rotation</button></div>
        </div>
      </div>
      <div class=\\"edit-tribe__section\\">
        <div class=\\"row\\">
          <div class=\\"col-12 col-md-3 pb-2 pb-md-0\\">
            <p class=\\"edit-tribe__subheading\\">Save changes</p>
            <p class=\\"text-muted\\">Save your changes before leaving this page.</p>
          </div>
          <div class=\\"col-12 col-md-9 d-flex align-items-center\\"><button class=\\"btn btn-block btn-primary\\">Save</button></div>
        </div>
      </div>
    </div>
  </div>
</fixedwidthwithnavbartemplate-stub>"
`;
